doc/dev/ragged_iter_domain_design_doc.md

```diff
@@ -0,0 +1,914 @@
+
+2. **Scheduling decisions**: Use `partition` directly for multi-GPU distribution strategies where work is split non-uniformly across devices
+
+3. **Merging back**: Use `merge` to convert ragged dimensions back to regular dimensions when needed (e.g., after expert processing, concatenate results back to merged form)
+
+The `partition` and `merge` operations enable ragged structures as shown in the examples below.
+
+#### Tensor Domain Structure
+
+A `TensorDomain` can contain a mix of regular `IterDomain` and `RaggedIterDomain` instances, representing tensors with both uniform and ragged dimensions. nvFuser supports **multiple RaggedIterDomains per TensorDomain** for load balancing scenarios.
+
+**Example 1: Single ragged dimension**
+```cpp
+// Batch of sequences with variable lengths
+auto batch = IrBuilder::create<IterDomain>(0, 3);
+auto ragged_seq = IrBuilder::create<RaggedIterDomain>({seq0, seq1, seq2});
+auto feature = IrBuilder::create<IterDomain>(0, 4);
+
+auto tensor_domain = IrBuilder::create<TensorDomain>({batch, ragged_seq, feature});
+```
+
+**Example 2: Multiple independent ragged dimensions**
+```cpp
+// Expert parallelism with load balancing
+// Different GPUs handle different numbers of experts (ragged)
+// Each expert processes different numbers of tokens (ragged)
+
+// Partition creates (batch_id, ragged_id) pairs
+auto [gpu_id, ragged_experts] = IterDomain::partition(expert_dim, gpu_offsets);
+// gpu_id: IterDomain with extent = number of GPUs
+// ragged_experts: RaggedIterDomain with variable experts per GPU
+
+auto [expert_id, ragged_tokens] = IterDomain::partition(token_dim, expert_offsets);
+// expert_id: IterDomain with extent = number of experts
+// ragged_tokens: RaggedIterDomain with variable tokens per expert
+
+auto hidden = IrBuilder::create<IterDomain>(0, 512);
+
+// TensorDomain must include BOTH the batch IterDomains AND the RaggedIterDomains
+auto tensor_domain = IrBuilder::create<TensorDomain>(
+    {gpu_id, ragged_experts, expert_id, ragged_tokens, hidden}
+);
+```
+
+The compilation pipeline detects ragged dimensions and routes to appropriate indexing and code generation paths.
+
+### 6.2 Core Abstractions
+
+#### RaggedIterDomain Class
+
+```cpp
+class RaggedIterDomain : public IterDomain {
+ public:
+  // Constructor
+  RaggedIterDomain(IrBuilderPasskey, std::vector<IterDomain*> nested_domains);
+
+  // Accessors
+  const std::vector<IterDomain*>& nestedDomains() const;
+  Val* extentForComponent(int64_t idx) const;
+
+  // This overrides IterDomain::extent and returns the total extent
+  Val* extent() const override;
+
+  // This overrides IterDomain::parallelize and calls nested_domains[i]->parallelize(pt) for all nested domains
+  void parallelize(ParallelType pt);
+
+  // Returns the offsets tensor (for compatibility with external APIs)
+  // Computes offsets from extents via cumulative sum
+  TensorView* offsets() const;
```

Comment:
For simplicity, in the first version, we can remove this and let GroupedMmaOp to compute offsets on demand. I don't expect much overhead at all given the size of the offset tensor and that prefixsum is done on GPU. The extent TensorView maps much more naturally to the extent of a RaggedIterDomain, so it should be much easier to implement.